staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go

```diff
@@ -0,0 +1,229 @@
+		// the sever was restarted.
+		if oldLease.Name == newLease.Name && *oldLease.Spec.HolderIdentity != *newLease.Spec.HolderIdentity {
+			h.withPeerAggDiscoveryCacheLock(func(cache map[string]*peerAggDiscoveryInfo) {
+				delete(cache, oldLease.Name)
+			})
+		}
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer to discovery cache", "serverID", newLease.Name)
+		}
+	}
+}
+
+func (h *peerProxyHandler) addPeerDiscoveryInfoToCache(serverIdentityLease *v1.Lease) error {
+	// Don't record our own discovery info.
+	if serverIdentityLease.Name == h.serverId {
+		return nil
+	}
+
+	var discoveryDoc *peerAggDiscoveryInfo
+	var ok bool
+	h.withPeerAggDiscoveryCacheRLock(func(cache map[string]*peerAggDiscoveryInfo) {
+		discoveryDoc, ok = cache[serverIdentityLease.Name]
+	})
+
+	if !ok ||
+		len(discoveryDoc.servedResources) == 0 ||
+		discoveryDoc.holderIdentity != *serverIdentityLease.Spec.HolderIdentity {
+		return h.fetchNewDiscoveryFor(serverIdentityLease.Name, *serverIdentityLease.Spec.HolderIdentity)
+	}
+
+	return nil
+}
+func isValidIdentityLease(obj interface{}, identityLeaseLabelSelector string) (*v1.Lease, bool) {
+	lease, ok := obj.(*v1.Lease)
+	if !ok {
+		klog.Error(fmt.Errorf("invalid lease object provided, received type: %T", obj))
+		return nil, false
+	}
+
+	parts := strings.Split(identityLeaseLabelSelector, "=")
+	if len(parts) != 2 {
+		klog.ErrorS(nil, "invalid identityLeaseLabelSelector provided, must be of the form key=value", "Received", identityLeaseLabelSelector)
+		return nil, false
+	}
+
+	identityLeaseComponentLabelKey := parts[0]
+	identityLeaseComponentLabelValue := parts[1]
+	if lease == nil || lease.Labels == nil ||
+		lease.Labels[identityLeaseComponentLabelKey] != identityLeaseComponentLabelValue {
+		klog.Error(fmt.Errorf("invalid lease object provided, must be an apiserver identity lease"))
+		return nil, false
+	}
+
+	if lease.Spec.HolderIdentity == nil {
+		klog.Error(fmt.Errorf("invalid lease object provided, missing holderIdentity in lease obj"))
+		return nil, false
+	}
+
+	return lease, true
+}
+
+func (h *peerProxyHandler) fetchNewDiscoveryFor(serverID string, holderIdentity string) error {
+	discoveryDoc := &peerAggDiscoveryInfo{
+		holderIdentity:  holderIdentity,
+		servedResources: make(map[schema.GroupVersion][]string),
+	}
+
+	hostport, err := h.hostportInfo(serverID)
+	if err != nil {
+		return fmt.Errorf("failed to get host port info from identity lease for server %s: %w", serverID, err)
+	}
+
+	klog.V(4).Infof("Proxying an agg-discovery call from %s to %s", h.serverId, serverID)
+	discoveryPaths := []string{"/api", "/apis"}
+	for _, path := range discoveryPaths {
+		discoveryResponse, err := h.aggregateDiscovery(path, hostport)
+		if err != nil {
+			klog.ErrorS(err, "error querying discovery endpoint", "path", path, "serverID", serverID)
+			continue
+		}
+
+		if discoveryResponse == nil {
```

Comment:
How does this happen? Is it not an error?

Reply:
it _is_ an error, so this is not needed, removed.